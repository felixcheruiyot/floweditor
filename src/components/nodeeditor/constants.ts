export const DEFAULT_OPERAND = '@input';
export const GROUPS_OPERAND = '@contact.groups';
export const SUBFLOW_OPERAND = '@child.run.status';
export const WEBHOOK_OPERAND = '@run.webhook';

export const DEFAULT_BODY: string = `{
    "flow": @(json(run.flow)),
    "path": @(json(run.path)),
    "results": @(json(run.results)),
    "input": @(json(run.input)),
    "channel": @(json(run.input.channel))
    "contact": @(json(contact))
}`;
