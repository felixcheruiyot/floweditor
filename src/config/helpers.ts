import { FlowTypes, FlowTypeVisibility, Operator, Type } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    return filterVisibility(operators, flowType);
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
