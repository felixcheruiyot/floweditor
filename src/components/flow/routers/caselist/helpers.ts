import { DraggableStyle } from 'react-beautiful-dnd';
import { CaseProps, DragCursor } from '~/components/flow/routers/caselist/CaseList';
import { Operators } from '~/config/interfaces';
import { getOperatorConfig } from '~/config/operatorConfigs';
import { createUUID } from '~/utils';

export const createEmptyCase = (): CaseProps => {
    const uuid = createUUID();
    const config = getOperatorConfig(Operators.has_any_word);
    return {
        uuid,
        kase: {
            uuid,
            type: config.type,
            arguments: [''],
            category_uuid: null
        },
        categoryName: '',
        valid: true
    };
};

export const getItemStyle = (
    draggableStyle: DraggableStyle,
    isDragging: boolean
): DraggableStyle => ({
    userSelect: 'none',
    outline: 'none',
    ...draggableStyle,
    // Overwriting default draggableStyle object from this point down
    ...(isDragging
        ? {
              background: '#f2f9fc',
              borderRadius: 4,
              opacity: 0.75,
              height: draggableStyle.height + 15
          }
        : {})
});

export const getListStyle = (isDraggingOver: boolean, single: boolean): { cursor: DragCursor } => {
    if (single) {
        return null;
    }

    return {
        cursor: isDraggingOver ? DragCursor.move : DragCursor.pointer
    };
};
