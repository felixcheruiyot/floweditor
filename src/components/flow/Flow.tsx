import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Canvas } from '~/components/canvas/Canvas';
import { CanvasDraggableProps } from '~/components/canvas/CanvasDraggable';
import Node from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import NodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from '~/components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from '~/config/ConfigProvider';
import { Exit, FlowDefinition } from '~/flowTypes';
import Plumber from '~/services/Plumber';
import editorState, { DragSelection, EditorState } from '~/store/editor';
import { RenderNode } from '~/store/flowContext';
import { detectLoops, getOrderedNodes } from '~/store/helpers';
import { NodeEditorSettings } from '~/store/nodeEditor';
import AppState from '~/store/state';
import {
    ConnectionEvent,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    mergeEditorState,
    MergeEditorState,
    NoParamsAC,
    OnConnectionDrag,
    onConnectionDrag,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    OnUpdateCanvasPositions,
    onUpdateCanvasPositions,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    updateSticky,
    UpdateSticky
} from '~/store/thunks';
import { contextTypes } from '~/testUtils';
import {
    ACTIVITY_INTERVAL,
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowStoreProps {
    editorState: Partial<EditorState>;
    mergeEditorState: MergeEditorState;

    definition: FlowDefinition;
    nodes: { [uuid: string]: RenderNode };
    dependencies: FlowDefinition[];
    nodeEditorSettings: NodeEditorSettings;

    ensureStartNode: EnsureStartNode;
    updateConnection: UpdateConnection;
    onOpenNodeEditor: OnOpenNodeEditor;
    onUpdateCanvasPositions: OnUpdateCanvasPositions;
    resetNodeEditingState: NoParamsAC;
    onConnectionDrag: OnConnectionDrag;
    updateSticky: UpdateSticky;
}

export interface Translations {
    [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowStoreProps, {}> {
    private Plumber: Plumber;
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        config: fakePropType
    };

    constructor(props: FlowStoreProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.config.debug) {
            window.fe = new Debug(props, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/, /^get/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.updateConnection(event.sourceId, event.targetId)
        );
        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.onConnectionDrag(event, this.context.config.flowType);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );
        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );
        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }

        this.props.mergeEditorState({
            containerOffset: { left: offset.left, top: offset.top + window.scrollY }
        });

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    public componentWillUpdate(prevProps: FlowStoreProps): void {
        if (
            prevProps.editorState.activityInterval === this.props.editorState.activityInterval &&
            this.props.editorState.activityInterval !== ACTIVITY_INTERVAL
        ) {
            this.props.mergeEditorState({ activityInterval: ACTIVITY_INTERVAL });
        }
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.resetNodeEditingState();
        const fromNodeUUID = event.sourceId.split(':')[0];
        try {
            detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
        } catch {
            return false;
        }
        return true;
    }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.editorState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = (this.ghost &&
                snapToGrid(
                    this.ghost.wrappedInstance.ele.offsetLeft,
                    this.ghost.wrappedInstance.ele.offsetTop
                )) || { left: 0, top: 0 };

            this.props.editorState.ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.onOpenNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        /* istanbul ignore next */
        document.removeEventListener('mousemove', (window as any).ghostListener);

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(): CanvasDraggableProps[] {
        return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
            return {
                uuid: renderNode.node.uuid,
                position: renderNode.ui.position,
                ele: (selected: boolean) => (
                    <Node
                        startingNode={idx === 0}
                        selected={selected}
                        key={renderNode.node.uuid}
                        data-spec={nodeSpecId}
                        nodeUUID={renderNode.node.uuid}
                        plumberMakeTarget={this.Plumber.makeTarget}
                        plumberRemove={this.Plumber.remove}
                        plumberRecalculate={this.Plumber.recalculate}
                        plumberMakeSource={this.Plumber.makeSource}
                        plumberConnectExit={this.Plumber.connectExit}
                        plumberUpdateClass={this.Plumber.updateClass}
                    />
                )
            };
        });
    }

    private getStickies(): CanvasDraggableProps[] {
        const stickyMap = this.props.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return {
                uuid,
                ele: (selected: boolean) => (
                    <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={selected} />
                ),
                position: stickyMap[uuid].position
            };
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.editorState.ghostNode) ? (
            <div style={{ position: 'absolute', display: 'block' }}>
                <Node
                    selected={false}
                    startingNode={false}
                    ref={this.ghostRef}
                    key={this.props.editorState.ghostNode.node.uuid}
                    data-spec={ghostNodeSpecId}
                    ghost={true}
                    nodeUUID={this.props.editorState.ghostNode.node.uuid}
                    plumberMakeTarget={this.Plumber.makeTarget}
                    plumberRemove={this.Plumber.remove}
                    plumberRecalculate={this.Plumber.recalculate}
                    plumberMakeSource={this.Plumber.makeSource}
                    plumberConnectExit={this.Plumber.connectExit}
                    plumberUpdateClass={this.Plumber.updateClass}
                />
            </div>
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(
            this.context.config.endpoints && this.context.config.endpoints.simulateStart
        )(<Simulator mergeEditorState={this.props.mergeEditorState} />);
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.nodeEditorSettings !== null)(
            <NodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.pageX - this.props.editorState.containerOffset.left - 100 + NODE_PADDING,
                event.pageY - this.props.editorState.containerOffset.top - NODE_PADDING * 2 - 40
            );

            this.props.updateSticky(createUUID(), {
                position: { left, top },
                title: STICKY_TITLE,
                body: STICKY_BODY
            });
        }
    }

    public render(): JSX.Element {
        const draggables = this.getStickies().concat(this.getNodes());

        return (
            <div onDoubleClick={this.onDoubleClick} ref={this.onRef}>
                <Canvas
                    onDragging={(uuids: string[]) => {
                        uuids.forEach((uuid: string) => {
                            if (uuid in this.props.nodes) {
                                this.props.nodes[uuid].node.exits.forEach((exit: Exit) => {
                                    if (exit.destination_uuid) {
                                        uuids.push(uuid + ':' + exit.uuid);
                                    }
                                });
                            }
                        });
                        this.Plumber.recalculateUUIDs(uuids);
                    }}
                    uuid={this.nodeContainerUUID}
                    dragActive={this.props.editorState.dragActive}
                    mergeEditorState={this.props.mergeEditorState}
                    draggables={draggables}
                    onUpdatePositions={this.props.onUpdateCanvasPositions}
                >
                    {this.getSimulator()}
                    {this.getDragNode()}
                    {this.getNodeEditor()}
                </Canvas>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({
    flowContext: { definition, dependencies, nodes },
    editorState,
    nodeEditor: { settings }
}: AppState) => {
    return {
        nodeEditorSettings: settings,
        definition,
        nodes,
        dependencies,
        editorState: editorState as Partial<EditorState>
    };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            mergeEditorState,
            ensureStartNode,
            resetNodeEditingState,
            onConnectionDrag,
            onOpenNodeEditor,
            onUpdateCanvasPositions,
            updateConnection,
            updateSticky
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Flow);
