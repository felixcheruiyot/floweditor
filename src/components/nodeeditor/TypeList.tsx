import * as React from 'react';
import Select from 'react-select';
import * as styles from '~/components/nodeeditor/TypeList.scss';
import { fakePropType } from '~/config/ConfigProvider';
import { filterTypeConfigs } from '~/config/helpers';
import { Type } from '~/config/interfaces';
import { configsToDisplay } from '~/config/typeConfigs';
import { large } from '~/utils/reactselect';

export interface TypeListProps {
    __className: string;
    initialType: Type;
    onChange(config: Type): void;
}

export interface TypeListState {
    config: Type;
}

export default class TypeList extends React.PureComponent<TypeListProps, TypeListState> {
    private typeConfigs: Type[];

    constructor(props: TypeListProps) {
        super(props);

        this.state = {
            config: this.props.initialType
        };

        this.handleChangeType = this.handleChangeType.bind(this);
    }

    public static contextTypes = {
        config: fakePropType
    };

    private handleChangeType(config: Type): void {
        this.setState(
            {
                config
            },
            () => this.props.onChange(config)
        );
    }

    private getTypeConfigs(): Type[] {
        if (this.typeConfigs === undefined) {
            this.typeConfigs = filterTypeConfigs(configsToDisplay, this.context.config.flowType);
        }
        return this.typeConfigs;
    }

    public render(): JSX.Element {
        return (
            <div className={`${this.props.__className} ${styles.typeList}`}>
                <p>When a contact arrives at this point in your flow...</p>
                <div>
                    <Select
                        styles={large}
                        value={this.state.config}
                        onChange={this.handleChangeType}
                        isSearchable={false}
                        isClearable={false}
                        getOptionValue={(option: Type) => option.type}
                        getOptionLabel={(option: Type) => option.description}
                        options={this.getTypeConfigs()}
                    />
                </div>
            </div>
        );
    }
}
