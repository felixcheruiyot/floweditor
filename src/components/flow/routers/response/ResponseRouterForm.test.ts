import { RouterFormProps } from '~/components/flow/props';
import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import ResponseRouterForm from '~/components/flow/routers/response/ResponseRouterForm';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Operators } from '~/config/interfaces';
import { Types } from '~/config/interfaces';
import { RouterTypes, SwitchRouter, WaitTypes } from '~/flowTypes';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, getRouterFormProps } from '~/testUtils/assetCreators';
import * as utils from '~/utils';
import { createUUID } from '~/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const { setup } = composeComponentTestUtils<RouterFormProps>(
    ResponseRouterForm,
    getRouterFormProps(
        createRenderNode({
            actions: [],
            exits: [],
            ui: { position: { left: 0, top: 0 }, type: Types.wait_for_response }
        })
    )
);

const otherExit = createUUID();
const redExit = createUUID();

const redCategory = createUUID();
const otherCategory = createUUID();

describe(ResponseRouterForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes', () => {
        const { wrapper } = setup(true, {
            nodeSettings: {
                $set: {
                    originalNode: createRenderNode({
                        actions: [],
                        exits: [
                            { destination_uuid: null, uuid: redExit },
                            { destination_uuid: null, uuid: otherExit }
                        ],
                        wait: { type: WaitTypes.msg },
                        router: {
                            type: RouterTypes.switch,
                            categories: [
                                {
                                    uuid: redCategory,
                                    name: 'Red'
                                },
                                {
                                    uuid: otherCategory,
                                    name: 'Other'
                                }
                            ],
                            cases: [
                                {
                                    uuid: createUUID(),
                                    type: Operators.has_any_word,
                                    arguments: ['red'],
                                    category_uuid: redCategory
                                }
                            ],
                            default_category_uuid: otherCategory,
                            result_name: 'Color'
                        } as SwitchRouter,
                        ui: {
                            position: { left: 0, top: 0 },
                            type: Types.wait_for_response
                        }
                    })
                }
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('initializes case config', () => {
        const dateCase = createUUID();
        const { wrapper } = setup(true, {
            nodeSettings: {
                $set: {
                    originalNode: createRenderNode({
                        actions: [],
                        exits: [
                            { destination_uuid: null, uuid: redExit },
                            { destination_uuid: null, uuid: otherExit }
                        ],
                        wait: { type: WaitTypes.msg },
                        router: {
                            type: RouterTypes.switch,
                            categories: [
                                {
                                    uuid: redCategory,
                                    name: 'Red'
                                },
                                {
                                    uuid: otherCategory,
                                    name: 'Other'
                                }
                            ],
                            cases: [
                                {
                                    uuid: dateCase,
                                    type: Operators.has_date_eq,
                                    arguments: ['@(datetime_add(today(), 5, "D"))'],
                                    category_uuid: redCategory
                                }
                            ],
                            default_category_uuid: otherCategory,
                            result_name: 'Color'
                        } as SwitchRouter,
                        ui: {
                            position: { left: 0, top: 0 },
                            type: Types.wait_for_response,
                            config: { router: { cases: { [dateCase]: { arguments: ['5'] } } } }
                        }
                    })
                }
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true, {
                $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
            });

            instance.handleUpdateTimeout(180);
            instance.handleUpdateResultName('Favorite Color');
            instance.handleCasesUpdated([
                {
                    uuid: createUUID(),
                    kase: { type: Operators.has_any_word, arguments: ['red'] },
                    categoryName: 'Red',
                    valid: true
                },
                {
                    uuid: createUUID(),
                    kase: { type: Operators.has_any_word, arguments: ['maroon'] },
                    categoryName: 'Red',
                    valid: true
                },
                {
                    uuid: createUUID(),
                    kase: { type: Operators.has_any_word, arguments: ['green'] },
                    categoryName: 'Green',
                    valid: true
                }
            ] as CaseProps[]);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).toHaveBeenCalled();
            expect(props.updateRouter).toMatchCallSnapshot();
        });

        it('should save save config for relative dates', () => {
            const { instance, props } = setup(true, {
                $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
            });

            instance.handleCasesUpdated([
                {
                    uuid: createUUID(),
                    kase: { type: Operators.has_date_gt, arguments: ['5'] },
                    categoryName: 'In the Zone',
                    valid: true
                }
            ] as CaseProps[]);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).toHaveBeenCalled();
            expect(props.updateRouter).toMatchCallSnapshot();
        });

        it('should cancel', () => {
            const { instance, props } = setup(true, {
                $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
            });
            instance.handleUpdateTimeout(180);
            instance.getButtons().secondary.onClick();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).not.toHaveBeenCalled();
        });
    });
});
