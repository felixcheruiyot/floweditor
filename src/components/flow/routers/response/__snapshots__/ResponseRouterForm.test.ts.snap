// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseRouterForm initializes 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  gutter={
    <TimeoutControl
      onChanged={[Function]}
    />
  }
  headerClass="wait_for_response"
  title="Wait for Response"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Wait for Response",
        "type": "wait_for_response",
        "visibility": Array [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div>
    If the message response...
  </div>
  <CaseList
    cases={
      Array [
        Object {
          "categoryName": "Red",
          "kase": Object {
            "arguments": Array [],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_any_word",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
          "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Color",
      }
    }
  />
</Dialog>
`;

exports[`ResponseRouterForm initializes case config 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  gutter={
    <TimeoutControl
      onChanged={[Function]}
    />
  }
  headerClass="wait_for_response"
  title="Wait for Response"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Wait for Response",
        "type": "wait_for_response",
        "visibility": Array [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div>
    If the message response...
  </div>
  <CaseList
    cases={
      Array [
        Object {
          "categoryName": "Red",
          "kase": Object {
            "arguments": Array [
              "5",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_date_eq",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Color",
      }
    }
  />
</Dialog>
`;

exports[`ResponseRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  gutter={
    <TimeoutControl
      onChanged={[Function]}
      timeout={0}
    />
  }
  headerClass="wait_for_response"
  title="Wait for Response"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Wait for Response",
        "type": "wait_for_response",
        "visibility": Array [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div>
    If the message response...
  </div>
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Result",
      }
    }
  />
</Dialog>
`;

exports[`ResponseRouterForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "red",
        ],
        "type": "has_any_word",
      },
      "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
      "valid": true,
    },
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "maroon",
        ],
        "type": "has_any_word",
      },
      "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
      "valid": true,
    },
    Object {
      "categoryName": "Green",
      "kase": Object {
        "arguments": Array [
          "green",
        ],
        "type": "has_any_word",
      },
      "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
      "valid": true,
    },
  ],
  "resultName": Object {
    "value": "Favorite Color",
  },
  "timeout": 180,
  "valid": true,
}
`;

exports[`ResponseRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
        },
        Object {
          "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        },
        Object {
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
        Object {
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "@input",
              "red",
            ],
            "category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "@input",
              "maroon",
            ],
            "category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "@input",
              "green",
            ],
            "category_uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "@run",
            ],
            "category_uuid": "f05fa107-03db-4315-a211-53da63603e6f",
            "omit_operand": true,
            "type": "has_wait_timed_out",
            "uuid": "141e0356-9fdd-47d3-a12d-16efc5b45276",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "6c22884b-0e35-4206-982e-18320691eda9",
            "name": "Red",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
          Object {
            "exit_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
            "name": "Green",
            "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
          },
          Object {
            "exit_uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
            "name": "Other",
            "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
          },
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "No Response",
            "uuid": "f05fa107-03db-4315-a211-53da63603e6f",
          },
        ],
        "default_category_uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        "result_name": "Favorite Color",
        "type": "switch",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "wait": Object {
        "timeout": 180,
        "type": "msg",
      },
    },
    "ui": Object {
      "config": Object {
        "router": Object {
          "cases": Object {},
          "operand": "@input",
        },
      },
      "position": null,
      "type": "wait_for_response",
    },
  },
]
`;

exports[`ResponseRouterForm updates should save save config for relative dates 1`] = `
Object {
  "cases": Array [
    Object {
      "categoryName": "In the Zone",
      "kase": Object {
        "arguments": Array [
          "5",
        ],
        "type": "has_date_gt",
      },
      "uuid": "e81b3f1a-0863-4418-8a72-9c9f8aa72b7c",
      "valid": true,
    },
  ],
  "resultName": Object {
    "value": "Result",
  },
  "timeout": 0,
  "valid": true,
}
`;

exports[`ResponseRouterForm updates should save save config for relative dates 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
        },
        Object {
          "uuid": "431978dd-6b18-4e25-86f1-11835491a9c9",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "@input",
              "@(datetime_add(today(), 5, \\"D\\"))",
            ],
            "category_uuid": "dd57500e-a8e8-4a98-8b5b-a33637523a8d",
            "type": "has_date_gt",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
            "name": "In the Zone",
            "uuid": "dd57500e-a8e8-4a98-8b5b-a33637523a8d",
          },
          Object {
            "exit_uuid": "431978dd-6b18-4e25-86f1-11835491a9c9",
            "name": "Other",
            "uuid": "3d8db86c-7279-4106-a388-3f143bd5623f",
          },
        ],
        "default_category_uuid": "3d8db86c-7279-4106-a388-3f143bd5623f",
        "result_name": "Result",
        "type": "switch",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "wait": Object {
        "type": "msg",
      },
    },
    "ui": Object {
      "config": Object {
        "router": Object {
          "cases": Object {
            "e81b3f1a-0863-4418-8a72-9c9f8aa72b7c": Object {
              "arguments": Array [
                "5",
              ],
            },
          },
          "operand": "@input",
        },
      },
      "position": null,
      "type": "wait_for_response",
    },
  },
]
`;
